Modules/BarcodeScanner/BarcodeScannerView.swift

```diff
@@ -0,0 +1,320 @@
+
+            if self.captureSession.canSetSessionPreset(.high) {
+                self.captureSession.sessionPreset = .high
+            }
+
+            guard let captureDevice = self.captureDevice else {
+                return DispatchQueue.main.async { [weak self] in
+                    self?.model.onError(.videoCapturingDeviceNotAvailable)
+                }
+            }
+
+            guard let captureDeviceInput = try? AVCaptureDeviceInput(device: captureDevice) else {
+                return DispatchQueue.main.async { [weak self] in
+                    self?.model.onError(.videoCapturingConfigurationFailed)
+                }
+            }
+
+            guard self.captureSession.canAddInput(captureDeviceInput) else { return }
+
+            self.captureSession.addInput(captureDeviceInput)
+            self.captureDeviceInput = captureDeviceInput
+
+            guard self.captureSession.canAddOutput(self.captureMetadataOutput) else {
+                return DispatchQueue.main.async { [weak self] in
+                    self?.model.onError(.videoCapturingConfigurationFailed)
+                }
+            }
+
+            self.captureSession.addOutput(self.captureMetadataOutput)
+            self.captureMetadataOutput.metadataObjectTypes = self.model.barcodeTypes.map { $0.rawValue }
+            self.captureMetadataOutput.setMetadataObjectsDelegate(self, queue: .main)
+            self.isCaptureSessionConfigured = true
+
+            DispatchQueue.main.async { [weak self] in
+                self?.setVideoOrientation(to: UIApplication.shared.statusBarOrientation)
+                self?.didChangeState()
+            }
+        }
+    }
+
+    public override func didChangeModel() {
+        super.didChangeModel()
+
+        captureSessionQueue.async { [weak self] in
+            guard let self = self, self.isCaptureSessionConfigured else { return }
+
+            self.captureSession.beginConfiguration()
+            defer { self.captureSession.commitConfiguration() }
+
+            self.captureMetadataOutput.metadataObjectTypes = self.model.barcodeTypes.map { $0.rawValue }
+        }
+    }
+
+    public override func willMove(toSuperview newSuperview: UIView?) {
+        switch (superview, newSuperview) {
+        case (.none, .some):
+            isScanningEnabled = true
+
+        case (.some, .none):
+            isScanningEnabled = false
+
+        default:
+            break
+        }
+    }
+
+    public override func layoutSubviews() {
+        super.layoutSubviews()
+
+        captureVideoPreviewLayer.frame = bounds
+    }
+
+    private func didChangeState() {
+        if isScanning, isScanningEnabled, !isCaptureSessionRunning {
+            self.captureSessionQueue.async { [weak self] in
+                guard let self = self else { return }
+
+                self.addObservers()
+                self.captureSession.startRunning()
+                self.isCaptureSessionRunning = self.captureSession.isRunning
```

Comment:
Why do you need this extra boolean type `isCaptureSessionRunning` as it always reflects `self.captureSession.isRunning`?

Reply:
The capture session is only accessed in the capture session queue so it is better to store the property to another variable to avoid race conditions with the queue. You could think of a situation where you request the isRunning value but the capture session is not fully configured at that time frame. Storing it into `isCaptureSessionRunning` is avoiding potential issues in consistency.